// mobx-angular/esm2015/lib/router-store.service.js
import { Injectable } from '@angular/core';
import { makeAutoObservable } from 'mobx';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class RouterStore {
    constructor(router, activatedRoute) {
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.url = '';
        this.routeSnapshot = null;
        makeAutoObservable(this);
        router.events
            .pipe(filter((event) => event instanceof NavigationEnd))
            .subscribe((e) => this.routeListener(e));
    }
    routeListener(event) {
        this.routeSnapshot = this.activatedRoute.snapshot;
        this.url = event.urlAfterRedirects;
    }
}
RouterStore.ɵprov = i0.ɵɵdefineInjectable({ factory: function RouterStore_Factory() { return new RouterStore(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i1.ActivatedRoute)); }, token: RouterStore, providedIn: "root" });
RouterStore.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
RouterStore.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute }
];
